tools/scripts/rebuildAnimationLibraryManifest.rb

```diff
@@ -0,0 +1,217 @@
+#
+# For usage, run ./rebuildAnimationLibraryManifest.rb --help
+#
+# The manifest has three important jobs:
+# 1. Provide the metadata used to render spritesheets as animations.
+# 2. Point to particular versions of the spritesheets on S3.
+# 3. Provide a precomputed search tree for quick animation searches by keyword.
+#
+# The expected output of this file is a new animationLibrary.json which is
+# consumed by the apps build.  You'll need to manually check this file into git.
+#
+# See also: The Animation Library Tech Spec (requires login):
+# https://docs.google.com/document/d/18-LVuvKd0jKTUiGo5GYReUWM5oFWCyKRyEQURJ5HCOM/edit
+#
+# TODO: Optimize: Read existing manifest and don't do full object reads for
+#       entries whose modify date hasn't changed.
+require 'aws-sdk'
+require 'optparse'
+require_relative '../../deployment'
+
+DEFAULT_S3_BUCKET = 'cdo-animation-library'
+DEFAULT_OUTPUT_FILE = 'apps/src/gamelab/animationLibrary.json'
+
+class ManifestBuilder
+  def initialize(options)
+    @options = options
+  end
+
+  #
+  # Main Entry Point
+  # See end of file for possible options, parsed from command line
+  #
+  def rebuild_animation_library_manifest
+    animations_by_name = {}
+
+    # Connect to S3 and get a listing of all objects in the animation library bucket
+    info 'Reading object list from S3...'
+    bucket = Aws::S3::Bucket.new(DEFAULT_S3_BUCKET)
+    bucket.objects.each do |object_summary|
+      animation_name = object_summary.key[/^[^.]+/]
+      extension = object_summary.key[/(?<=\.)\w+$/]
+      verbose <<-EOS.unindent
+        #{bold object_summary.key}
+          #{object_summary.last_modified} | #{object_summary.size}
+      EOS
+      # Push into animations collection
+      animations_by_name[animation_name] ||= {}
+      animations_by_name[animation_name][extension] = object_summary
+    end
+
+    # TODO: Validate that keys are unique
+
+    info "Found #{animations_by_name.size} animations."
+
+    animation_metadata_by_name = {}
+    info "Building animation metadata..."
+    animations_by_name.each do |name, objects|
+      # TODO: Validate that every JSON is paired with a PNG and vice-versa
+      # Actually download the JSON from S3
+      json_response = objects['json'].get
+      metadata = JSON.parse(json_response.body.read)
+      # TODO: Validate metadata, ensure it has everything it needs
+      # Record target version in the metadata, so environments (and projects)
+      # consistently reference the version they originally imported.
+      metadata['version'] = objects['png'].object.version_id
+
+      # Generate appropriate sourceUrl pointing to the animation library API
+      metadata['sourceUrl'] = "/api/v1/animation-library/#{metadata['version']}/#{name}.png"
+
+      # Populate sourceSize if not already present
+      unless metadata.key?('sourceSize')
+        png_body = objects['png'].object.get.body.read
+        metadata['sourceSize'] = dimensions_from_png(png_body)
```

Comment:
This part of the script can get slow - getting the object versions is a [`HEAD Object`](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectHEAD.html) operation, I think, which is bad enough, but to read the source sizes we actually have to download the PNGs themselves.  While I don't think this traffic is going to hurt us for an occasional script run, I'd like to do a little optimization and use the existing manifest to skip any images that haven't changed.